daemon/cmd/daemon_main.go

```diff
@@ -1258,12 +1259,30 @@ func initEnv(cmd *cobra.Command) {
 		}
 	}
 
-	if option.Config.EnableIPSec && option.Config.Tunnel == option.TunnelDisabled && option.Config.EncryptInterface == "" {
-		link, err := linuxdatapath.NodeDeviceNameWithDefaultRoute()
-		if err != nil {
-			log.WithError(err).Fatal("Ipsec default interface lookup failed, consider \"encrypt-interface\" to manually configure interface.")
+	if option.Config.EnableIPSec && option.Config.Tunnel == option.TunnelDisabled && len(option.Config.EncryptInterface) == 0 {
+		// IPAMENI mode supports multiple network facing interfaces that
+		// will all need Encrypt logic applied in order to decrypt any
+		// received encrypted packets. This logic will attach to all
+		// !veth devices.
+		if option.Config.IPAM == ipamOption.IPAMENI {
+			interfaces := option.Config.EncryptInterface
+
+			if links, err := netlink.LinkList(); err == nil {
+				for _, link := range links {
+					use, err := ethtool.IsNotVirtualDriver(link.Attrs().Name)
+					if err == nil && use {
+						interfaces = append(interfaces, link.Attrs().Name)
+					}
+				}
+				option.Config.EncryptInterface = interfaces
```

Comment:
We are overriding user-specified config option here, no? I wonder if this can lead to confusion. At the very least, it'll be good to document this?

Reply:
If the user set an interface we skip this entire block.
```
        if option.Config.EnableIPSec && option.Config.Tunnel == option.TunnelDisabled && len(option.Config.EncryptInterface) == 0 {
                // IPAMENI mode supports multiple network facing interfaces that
                // will all need Encrypt logic applied in order to decrypt any
                // received encrypted packets. This logic will attach to all
                // !veth devices.
                if option.Config.IPAM == ipamOption.IPAMENI {
                        interfaces := option.Config.EncryptInterface

                        if links, err := netlink.LinkList(); err == nil {
                                for _, link := range links {
                                        use, err := ethtool.IsNotVirtualDriver(link.Attrs().Name)
                                        if err == nil && use {
                                                interfaces = append(interfaces, link.Attrs().Name)
                                        }
                                }
                                option.Config.EncryptInterface = interfaces
                        }
                } else {
                        link, err := linuxdatapath.NodeDeviceNameWithDefaultRoute()
                        if err != nil {
                                log.WithError(err).Fatal("Ipsec default interface lookup failed, consider \"encrypt-interface\" to manually configure interface.")
                        }
                        option.Config.EncryptInterface = append(option.Config.EncryptInterface, link)
                }
        }
```

Reply:
In the ENI mode, users could specify incorrect interface which can lead to packets not getting decrypted. Any reason why we don't want to enable auto-discovery in such environments regardless of the user-specified config?

Reply:
I think if the user adds interfaces we should do what they tell us. It gives us an escape route if this breaks for whatever reason.